kramer/src/main/kotlin/kramer/MavenRepo.kt

```diff
@@ -0,0 +1,331 @@
+          }
+          .asMap()
+          .entries
+          .asFlow()
+          .flatMapMerge(concurrency = threadCount) { (path, snippets) ->
+            flow {
+              val android = snippets.map { it.second }.reduce { b, acc -> b || acc }
+              val androidHeader = if (android) ANDROID_LOAD_HEADER else ""
+              val mavenRulesRepo = "maven_repository_rules"
+              val content = snippets.joinToString(
+                  "\n",
+                  prefix = HEADER + androidHeader + RAW_LOAD_HEADER_TEMPLATE.format(mavenRulesRepo)
+                ) { it.first }
+              Files.createDirectories(path.parent)
+              Files.write(path, content.lines())
+              emit(path to snippets)
+            }
+          }
+          .collect()
+
+        // Check for, and handle, any errors or mis-specifications.
+        if (declaredArtifactSlugs.size != repoConfig.artifacts.size) {
+          exit = 1
+          handleDuplicateArtifacts(repoConfig)
+        }
+        if (unresolved.isNotEmpty()) {
+          exit = 1
+          handleUnresolvedArtifacts(unresolved)
+        }
+        with(seen.filterKeys { it !in declaredArtifactSlugs }) {
+          if (isNotEmpty()) {
+            exit = 1
+            handleMissingArtifacts(this)
+          }
+        }
+      }
+    }
+    kontext.out {
+      "Resolved $count artifacts with $threadCount threads in ${benchmark / 1000.0} seconds"
+    }
+    if (exit != 0) throw ProgramResult(exit)
+  }
+}
+
+/**
+ * For a given resolved artifact, prepare the list of its dependencies, excluding unused scopes,
+ * explicitly excluded deps, and fixing and formatting.
+ */
+private fun prepareDependencies(
+  resolved: ResolvedArtifact,
+  config: ArtifactConfig,
+  seen: ConcurrentHashMap<String, MavenRepo.IndexEntry>,
+  repoConfig: RepoConfig
+): List<String> {
+  val substitutes = repoConfig.targetSubstitutes.getOrElse(resolved.groupId) { mapOf() }
+  return resolved.model.dependencies
+    .filter { dep -> dep.scope in acceptedScopes }
+    .filter { dep -> "${dep.groupId}:${dep.artifactId}" !in config.exclude }
+    .onEach { dep ->
+      // Cache for later validation
+      val entry = seen.getOrPut("${dep.groupId}:${dep.artifactId}") { MavenRepo.IndexEntry() }
+      entry.versions.add(dep.version)
+      entry.dependants.add(resolved.coordinate)
+    }
+    .map { dep ->
+      val bazelPackage = "@${repoConfig.name}//${dep.groupPath}"
+      val leafPackage = bazelPackage.split("/").last()
+      val prefix = "$bazelPackage:"
+      "$prefix${dep.target}"
+        .let { d ->
+          substitutes.getOrElse(d) { d }
+        }
+        .let { d ->
+          when {
+            resolved.groupId == dep.groupId -> d.removePrefix(bazelPackage)
+            leafPackage == dep.target -> bazelPackage
+            else -> d
+          }
+        }
+    }
+    .sorted()
+}
+
+@Throws(IOException::class)
+private fun extractPackageFromManifest(fromZip: URI): String? {
```

Comment:
Move to an aar related file? Seems weird to keep it main.

Reply:
Thought about that. Was on the fence because a file with one function seemed a bit off, but I concur. I'll make that happen.

Reply:
Done.